src/auth.rs

```diff
@@ -100,60 +167,156 @@ impl<'a> Auth<'a> {
-                params: self.rest.opts.auth_params.clone(),
-            });
-        } else if let Some(ref key) = self.rest.opts.key {
-            builder = builder.key(key.clone());
-        } else if let Some(ref token) = self.rest.opts.token {
-            builder = builder.token(token.clone());
-        }
+    pub async fn request_token(
+        &self,
+        params: &TokenParams,
+        options: &AuthOptions,
+    ) -> Result<TokenDetails> {
+        let token = options
+            .token
+            .as_ref()
+            .ok_or_else(|| error!(40171, "no means provided to renew auth token"))?;
+
+        let mut details = match token {
+            TokenSource::TokenDetails(token) => Ok(token.clone()),
+            TokenSource::TokenRequest(r) => self.exchange(r).await,
+            TokenSource::Callback(f) => match f.token(params).await {
+                Ok(token) => token.into_details(self).await,
+                Err(e) => Err(e),
+            },
+            TokenSource::Key(k) => self.exchange(&params.sign(k)?).await,
+            TokenSource::Url(url) => self.request_url(url).await,
+        };
 
-        if let Some(params) = &self.rest.opts.default_token_params {
-            builder = builder.params(params.clone());
+        if matches!(token, TokenSource::Callback(_) | TokenSource::Url(_)) {
+            if let Err(ref mut err) = details {
+                // Normalise auth error according to RSA4e.
+                if err.code == 40000 {
+                    err.code = 40170;
+                    err.status_code = Some(401);
+                }
+            };
         }
 
-        if let Some(client_id) = &self.rest.opts.client_id {
-            builder = builder.client_id(client_id);
+        let details = details?;
+
+        // Reject tokens with size greater than 128KiB (RSA4f).
+        if details.token.len() > MAX_TOKEN_LENGTH {
+            return Err(error!(
+                40170,
+                format!(
+                    "Token string exceeded max permitted length (was {} bytes)",
+                    details.token.len()
+                ),
+                401
+            ));
         }
 
-        builder
+        Ok(details)
     }
 
     /// Set the Authorization header in the given request.
-    pub async fn with_auth_headers(&self, req: &mut reqwest::Request) -> Result<()> {
-        if let Some(ref key) = self.rest.opts.key {
-            if !self.rest.opts.use_token_auth {
-                return Self::set_basic_auth(req, key);
-            }
+    pub(crate) async fn with_auth_headers(&self, req: &mut reqwest::Request) -> Result<()> {
+        if let TokenSource::Key(k) = &self.inner().opts.token {
+            return Self::set_basic_auth(req, k);
         }
 
-        let res = self.request_token().send().await?;
+        let options = AuthOptions {
+            token: Some(self.inner().opts.token.clone()),
+            ..Default::default()
```

Comment:
What I mean is, let's say I have an auth server at `https://example.com/auth` that returns Ably tokens, but I require the request to be a `POST` request with some special header set, then as part of the client options, I can set `auth_method` and `auth_headers` (see [here](https://github.com/ably/ably-rust/blob/main/src/options.rs#L28-L38)), which I expect to be honoured when the SDK makes requests to my auth server.

So here, I'm asking whether `auth_method`, `auth_headers`, and `auth_params` from the `ClientOptions` ought to be set on `options` so that the request below uses those values?

Reply:
They're still there here https://github.com/ably/ably-rust/blob/c57a1e24804a591a303723efa58092d824e3d67c/src/auth.rs#L84

There's just no builder methods as of yet.

Reply:
Ok so really, what I'm asking is, should this be something like:

```rust
let options = AuthOptions {
  token: Some(self.inner().opts.token.clone()),
  headers: self.inner().opts.auth_headers,
  method: self.inner().opts.auth_method,
  params: self.inner().opts.auth_params,
}
```